File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case18_cep/Case02_IndividualCEP.java
package com.yw.flink.example.javacases.case18_cep;

import org.apache.flink.cep.CEP;
import org.apache.flink.cep.PatternSelectFunction;
import org.apache.flink.cep.PatternStream;
import org.apache.flink.cep.pattern.Pattern;
import org.apache.flink.cep.pattern.conditions.IterativeCondition;
import org.apache.flink.cep.pattern.conditions.SimpleCondition;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

import java.util.List;

/**
 * 简单事件定义模式 ： 设置量词和条件
 */
public class Case02_IndividualCEP {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //1.定义事件流
        DataStreamSource<String> ds = env.socketTextStream("nc_server", 9999);

        //2.定义Pattern模式
        Pattern<String, String> pattern = Pattern.<String>begin("first")
                .where(new SimpleCondition<String>() {
                    @Override
                    public boolean filter(String s) throws Exception {
                        return s.startsWith("a");
                    }
                }).oneOrMore()
                .until(new IterativeCondition<String>() {
                    @Override
                    public boolean filter(String s, Context<String> context) throws Exception {
                        return s.equals("end");
                    }
                });

        //3.模式作用到数据流中
        PatternStream<String> patternStream = CEP.pattern(ds, pattern).inProcessingTime();

        //4.获取匹配结果
        SingleOutputStreamOperator<String> result = patternStream
                .select((PatternSelectFunction<String, String>) map -> {
                    List<String> first = map.get("first");
                    StringBuilder builder = new StringBuilder();
                    for (String s : first) {
                        builder.append(s).append("-");
                    }
                    return builder.substring(0, builder.length() - 1);
                });

        result.print();
        env.execute();
    }
}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case18_cep/Case02_IndividualCEP.scala
package com.yw.flink.example.scalacases.case18_cep

import org.apache.flink.cep.PatternSelectFunction
import org.apache.flink.cep.scala.pattern.Pattern
import org.apache.flink.cep.scala.{CEP, PatternStream}
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment

import java.util

/**
  * 简单事件定义模式 ： 设置量词和条件
  *
  */
object Case02_IndividualCEP {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.streaming.api.scala._

    //1.定义事件流
    val ds: DataStream[String] = env.socketTextStream("nc_server", 9999)

    //2.定义匹配规则
    val pattern: Pattern[String, String] = Pattern.begin[String]("first").where(_.startsWith("a"))
      .oneOrMore
      .until(_.equals("end"))

    //3.将规则应用到事件流上
    val patternStream: PatternStream[String] = CEP.pattern(ds, pattern).inProcessingTime()

    //4.获取匹配结果
    val result: DataStream[String] = patternStream.select(new PatternSelectFunction[String, String] {
      override def select(map: util.Map[String, util.List[String]]): String = {
        val list: util.List[String] = map.get("first")
        import scala.collection.JavaConverters._
        list.asScala.mkString("-")
      }
    })

    result.print()
    env.execute()
  }

}
